mpa/modules/models/classifiers/sam_classifier.py

```diff
@@ -257,11 +257,17 @@ def extract_feat(self, img):
            Overriding for OpenVINO export with features
         """
         x = self.backbone(img)
+        # For Global Backbones (det/seg/etc..),
+        # In case of tuple or list, only the feat of the last layer is used.
+        if isinstance(x, (tuple, list)):
+            x = x[-1]
```

Comment:
Hi! @negvet I need your advice so leave a comment.. :)
I have implemented the current public backbone (detection or segmentation) for use by attaching it to the classification model. -> Ex: {mmdet backbone + classifier}
The detection backbone basically emits `multiple outputs` and returns a `tuple or list`, and in order to use it for classification, the last output is subtracted and passed in this way now. `x = x[-1]`

However, recently, as the `ActivationMapHook` on the MPA side was changed to `ReciproCAMHook`, there was a problem in the infer process (Only `classification` + when the output of backbone is a `tuple or list` -> I think this currently has no test code.)
https://github.com/openvinotoolkit/model_preparation_algorithm/blob/c1e801f24595a95594be3520ecdf97efdd73263d/mpa/modules/hooks/recording_forward_hooks.py#L266 
I think the conflicting part is here, and when the output is a `tuple or list`, the `feature_map` at `index 0` (default value) is taken out and used. 

It seems that there is a conflict in the calculation because the head is different from the known size, and it has been confirmed that it works when `fpn_dix=-1`

I'm not sure if my explanation is clear, but summary is that there's a problem when the `feature_map` exported from backbone is a `tuple or list` pass to classification head, and it seems to be caused by different indices. (My guess so far is that...)

Can you give me an idea or solution to this problem or what I am doing wrong or what needs to be fixed?

Reply:
Hi Harim

Thanks for a notice!

You are right.

I do not know why `fpn_idx` defaults to 0. It was 0 before and I did not update it (my bad, and yes, imo we do not have tests for num_backbone_outputs > 1 for cls). `fpn_idx` is the index of the backbone output (while increasing the index - it approaching the head). By default we should take the last layer output for explanation perposes. So `fpn_idx` should default to -1. 

`fpn_idx=-1` is a good default for `ActivationMapHook`. But this is not a must to be. In practice we might want to see `fpn_idx=0` for example (`ActivationMapHook` will work with any `fpn_idx` from possible idx range). So we should keep this argument. 

`ReciproCAMHook`, on the other hand, process the output from the backbone and forward it through the cls head. Then it makes a big difference which `fpn_idx` we pick (we need to make sure that cls head can get it as input). 
Assuming that the cls heads only works with the higher fpn level (`fpn_idx=-1`), then for `ReciproCAMHook` we also need to assign `fpn_idx=-1`. 

> Assuming that the cls heads only works with the higher fpn level

If this is true, then we actually need to hardcode it to be `x = x[-1]`. Because no other options are available (in contrast to `ActivationMapHook`).

Can assume that there might be other advanced options: (1) cls head work with NOT the higher fpn level, (2) cls head can take as input all fpn levels. These more advanced use cases are not supported by `ReciproCAMHook`. It is a good to raise an error in this case.

Hope it was clear.

Should I open a PR to update `fpn_idx` default values or you have it already somewhere?

Reply:
@harimkang @negvet , In this situation, we should select `x=x[-1]` for the classifier and `fpn_idx=-1` for the ReciproCAMHook. And actually, IMO, ReciproCAMHook can only take the last backbone layer because it needs to connect with `cls_head` like as below.
![image](https://user-images.githubusercontent.com/80735344/209489799-a50df81d-4c26-4f41-a354-12f2b0db184c.png)

And I think changing the cls_head (advanced option 1, 2) is inappropriate. Since CAM assumes that all layers included in the model were pre-trained, changing the size of cls_head might not work as we anticipated. 
(i.e. Assume that the model was trained on 1,024 size cls_head and we want to use a 2,048 feature map. In this case, we can change the size of cls_head to 1,024 --> 2,048 to resolve mis-size issue. However, in this case, cls_head will not work because it was trained on 1,024)

So, let's use `fpn_idx=-1` as a solution. And maybe we need to extract the saved tensor and avoid re-forward to make the tensor output.